File: src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from functools import partial

from example_interfaces.srv import AddTwoInts


class AddTwoIntsClientNode(Node):

    def __init__(self):
        super().__init__("add_two_ints_client")
        self.ask_call_loop()

    def ask_call_loop(self):
        try:
            print("Insert a: ", end="")
            self.a_ = int(input())
            print("Insert b: ", end="")
            self.b_ = int(input())

            self.call_two_ints_server(self.a_, self.b_)
        except Exception as e:
            print(e)

    def call_two_ints_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")

        while(not client.wait_for_service(1.0)):
            self.get_logger().warn("Waiting for server add_two_ints to be up")

        '''Service is up now'''
        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        future = client.call_async(request)
        future.add_done_callback(partial(self.callback_add_two_ints, a=a, b=b))

    def callback_add_two_ints(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(
                "Response is {0} for request \"{1} + {2}\"".format(response.sum, a, b))

            self.ask_call_loop()

        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))


def main(args=None):
    
    rclpy.init(args=args)  # first line to be written in any ROS2 .py node
    node = AddTwoIntsClientNode()
    rclpy.spin(node)  # allow node to continue to be alive
    rclpy.shutdown()  # last line of any ROS2 .py node

if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

using std::vector;
using std::thread;
using std::bind;

using example_interfaces::srv::AddTwoInts;

class AddTwoIntsClientNode : public rclcpp::Node
{
public:
    AddTwoIntsClientNode() : Node("add_two_ints_client")
    {
        threads.push_back(thread(bind(&AddTwoIntsClientNode::callAddTwoInts, this)));
    }

private:
    void callAddTwoInts()
    {
        int a, b;

        std::cout << "Insert a: "; std::cin >> a;
        std::cout << "Insert b: "; std::cin >> b;

        rclcpp::Client<AddTwoInts>::SharedPtr client_ = this->create_client<AddTwoInts>("add_two_ints");
        while(!client_->wait_for_service(std::chrono::seconds(1)))
            RCLCPP_WARN(this->get_logger(), "Waiting for Two Ints server to be up");

        auto request = std::make_shared<AddTwoInts::Request>();
        request->a = a;
        request->b = b;

        auto future = client_->async_send_request(request);

        try{
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "Response is %d for sum %d + %d", response->sum, request->a, request->b);

            threads.push_back(thread(bind(&AddTwoIntsClientNode::callAddTwoInts, this)));
        }catch(const std::exception &e){
            RCLCPP_ERROR(this->get_logger(), "Error while calling server for sum %d + %d", request->a, request->b);
        }
    }

    vector<thread> threads;

};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    //shared pointer, you don't need to delete manually the instance when object is not referenced (like GC in Java)
    auto node = std::make_shared<AddTwoIntsClientNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}